### PR TITLE
 added to type mismatch reporting output of tree, which can't be typed

### PR DESCRIPTION
**Pull request by rssh**
added to type mismatch reporting output of tree, which can't be typed in -explain mode.
Because in macros, tree in sourcecode in position and a processed tree can be very different.

### CODE REVIEW

**Comment by SethTisue** [comment]
Looks plausible, but needs test coverage, yes? (In part because a test or two would help illustrate the purpose of the PR.)

**Review by dwijnand — CHANGES_REQUESTED** [review]
I agree, the intent seems reasonable to me, to help understand what went wrong in the expanded macro code.  But we'd want to see some examples as tests to review and maintain this enhancement.

File: compiler/src/dotty/tools/dotc/reporting/messages.scala
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/reporting/messages.scala:288]: To avoid diverging from the parent's definition of explain, let's reuse it here:

```suggestion
      treeStr + "\n" + super.explain
```

**Comment by rssh** [comment]
Just added tests.   will rebase and squash now.
Short description of test-case here:

```scala
package t12717

import scala.quoted._

object A:

   def foo(x:Int): Int = ???

   def bar(x:String): String = ???


object X:

   inline def doSomething[T](inline x:T):Any = ${
      doSomethingImpl('x)
   }

   def doSomethingImpl[T:Type](x:Expr[T])(using Quotes):Expr[Any] =
     import quotes.reflect._
     val aTerm = '{A}.asTerm
     val xBar = Apply(Select.unique(aTerm,"bar"),List(x.asTerm))
     Apply(Select.unique(aTerm,"foo"), List(xBar)).asExpr
``` scala

And usage:
```
package t12717

object Test:

   val x = X.doSomething("XXX") // error

```

The output is:
```
-- [E007] Type Mismatch Error: tests/neg-custom-args/hidden-type-errors/Test.scala:6:24 --------------------------------
6 |   val x = X.doSomething("XXX") // error
  |           ^^^^^^^^^^^^^^^^^^^^
  |           Found:    String
  |           Required: Int
  | This location contains code that was inlined from Test.scala:6

Explanation
===========

Tree: t12717.A.bar("XXX")

I tried to show that
  String
conforms to
  Int
but the comparison trace ended with `false`:


  ==> String  <:  Int
    ==> String  <:  Int (recurring)
      ==> String  <:  Int (recurring)
      <== String  <:  Int (recurring) = false
    <== String  <:  Int (recurring) = false
  <== String  <:  Int = false

The tests were made under the empty constraint

1 error found
```

Without 
```
Tree: t12717.A.bar("XXX")
```

it's hard to understand, why the type checking of macro result has failed.

**Comment by rssh** [comment]
btw, also I'm not sure that I added the test in the right place:  looks like a neg-custom-args check only the existence of error, not an exact output of the compiler.  Maybe we need a new category for such test types (?)

**Review by dwijnand — APPROVED** [review]
Seems reasonable to me.  What I see missing is that `Int` is expected because that tree is being passed to `A.foo`, which is still only understandable from reading the macro.  So it almost seems like a half improvement, but that's still an improvement over the status quo.

I wouldn't mind another maintainer weighing in with or against me, before merging.

**Review by OlivierBlanvillain — APPROVED** [review]

**Pull request merged**
